File: elixir/largest-series-product/largest_series_product.exs
defmodule Series do
  require Integer

  @doc """
  Finds the largest product of a given number of consecutive numbers in a given string of numbers.
  """
  @spec largest_product(String.t, non_neg_integer) :: non_neg_integer
  def largest_product(_, 0), do: 1
  def largest_product(number_string, size) do
    if size < 0, do: raise ArgumentError
    if String.length(number_string) < size, do: raise ArgumentError

    number_string
    |> String.codepoints
    |> Enum.map(&String.to_integer/1)
    |> Enum.chunk(size, 1)
    |> Enum.map(&(Enum.reduce(&1, 1, fn x, acc -> acc * x end)))
    |> Enum.max
  end
end

File: javascript/largest-series-product/largest-series-product.js
var Series = function(series) {
  if (!series.match(/^\d*$/g)) { throw new Error('Invalid input.'); }
  this.series = [...series];
};

Series.prototype.largestProduct = function(size) {
  if (size > this.series.length) { throw new Error('Slice size is too big.'); }
  if (this.series.length == 0 || size == 0) { return 1; }
  if (size < 0) { throw new Error('Invalid input.'); }
  let out = 0;
  for (var x=0; x<=this.series.length-size; x++) {
    let slice = this.series.slice(x, x+size);
    let product = slice.reduce((p, c) => p * c, 1);
    out = (product > out) ? product : out;
  }
  return out;
};

module.exports = Series;
